app/src/main/java/org/schabi/newpipe/player/helper/PlayerHelper.java

```diff
@@ -163,13 +169,39 @@ public static String resizeTypeOf(@NonNull final Context context,
     @NonNull
     public static String cacheKeyOf(@NonNull final StreamInfo info,
                                     @NonNull final VideoStream video) {
-        return info.getUrl() + video.getResolution() + video.getFormat().getName();
+        String cacheKey = info.getUrl();
+        final String resolution = video.getResolution();
+        if (!resolution.equals(RESOLUTION_UNKNOWN)) {
+            cacheKey += resolution;
+        } else {
+            cacheKey += String.valueOf(cachedVideoStreamsWithResolutionUnknown);
+            cachedVideoStreamsWithResolutionUnknown += 1;
```

Comment:
It doesn't seem like a good idea to me to use a static counter to differentiate between different cached video/audio streams. Isn't there any other information that the extractor could provide that allows differentiating between streams? Otherwise it's just better to disable cache for these streams, which is probably what you are trying to achieve by making a different key each time, but this way the cache is not really disabled for these streams.

Reply:
Yes, I am agree with you.

What do you think of this code instead?

- for video streams:
  ```java
  @NonNull
  public static String cacheKeyOf(@NonNull final StreamInfo info,
                                  @NonNull final VideoStream videoStream) {
      String cacheKey = info.getUrl() + " " + videoStream.getId();

      final String resolution = videoStream.getResolution();
      final MediaFormat mediaFormat = videoStream.getFormat();
      if (resolution.equals(RESOLUTION_UNKNOWN) && mediaFormat == null) {
          final String content = videoStream.getContent();
          cacheKey += " " + content.substring(20 * content.length() / 100);
      } else {
          if (mediaFormat != null) {
              cacheKey += " " + videoStream.getFormat().getName();
          }
          if (!resolution.equals(RESOLUTION_UNKNOWN)) {
              cacheKey += " " + resolution;
          }
      }

      return cacheKey;
  }
  ```

- for audio streams:
  ```java
  @NonNull
  public static String cacheKeyOf(@NonNull final StreamInfo info,
                                  @NonNull final AudioStream audioStream) {
      String cacheKey = info.getUrl() + " " + audioStream.getId();
  
      final int averageBitrate = audioStream.getAverageBitrate();
      final MediaFormat mediaFormat = audioStream.getFormat();
      if (averageBitrate == UNKNOWN_BITRATE && mediaFormat == null) {
          final String content = audioStream.getContent();
          cacheKey += " " + content.substring(20 * content.length() / 100);
      } else {
          if (mediaFormat != null) {
              cacheKey += " " + audioStream.getFormat().getName();
          }
          if (averageBitrate != UNKNOWN_BITRATE) {
              cacheKey += " " + averageBitrate;
          }
      }
  
      return cacheKey;
  }
  ```